### PR TITLE
[lingui__react]: Add i18nHash to withI18n

### PR DESCRIPTION
**Pull request by robertcoopercode**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
  - https://github.com/lingui/js-lingui/blob/afc05d57336ab38f648bcfe52711f80d230fa47d/packages/react/src/I18nProvider.js#L22
  - https://github.com/lingui/js-lingui/blob/afc05d57336ab38f648bcfe52711f80d230fa47d/packages/react/src/I18nProvider.js#L70
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

### CODE REVIEW

**Comment by huan086** [comment]
Note to self:

[I18nProvider](https://github.com/lingui/js-lingui/blob/v2.9.1/packages/react/src/I18nProvider.js#L42) stores null in i18nHash. 

When [withHash](https://github.com/lingui/js-lingui/blob/v2.9.1/packages/react/src/I18n.js#L41) is true, i18nHash may be null or string. When withHash is false, i18nHash is undefined

**Review by huan086 — APPROVED** [review]

**Comment by robertcoopercode** [comment]
Thanks for the quick review @huan086

File: types/lingui__react/withI18n.d.ts [hunk 1]
```diff
@@ -11,6 +11,7 @@ export interface withI18nOptions {
 
 export interface withI18nProps {
     i18n: I18n;
+    i18nHash: string | null | undefined;
```
Comment (amcasey) [inline types/lingui__react/withI18n.d.ts]: Was this intended to be optional?  Right now, it has to be provided, but can be null or undefined (which may be intentional).

**Review by amcasey — CHANGES_REQUESTED** [review]
Minor question.  Happy to merge as-is.

File: types/lingui__react/withI18n.d.ts (hunk 1 above)
Comment (robertcoopercode) [reply types/lingui__react/withI18n.d.ts]: No, it's the `i18nHash` prop is always injected into a component and therefore is a non-optional type:

https://github.com/lingui/js-lingui/blob/7f4cc1546546ff3b0ce63ec4316ce6d0539c670a/packages/react/src/withI18n.js#L65

As @huan086 mentioned: When withHash is true, i18nHash may be null or string. When withHash is false, i18nHash is undefined

**Review by amcasey — APPROVED** [review]
Sorry - lost track of this.

**Pull request merged**